pootle/static/js/editor/UnitList.js

```diff
@@ -0,0 +1,294 @@
+            }
+          }
+          this.unit.contextData = out;
+          this.handleContextDataChange(uid);
+        },
+        this.error
+      );
+    } else {
+      this.handleContextDataChange(uid);
+    }
+  }
+
+  gotoPrev() {
+    const idx = Math.max(0, this.idx - 1);
+    return this.goto(this.uids[idx]);
+  }
+
+  gotoNext() {
+    const idx = Math.min(this.uids.length, this.idx + 1);
+    return this.goto(this.uids[idx]);
+  }
+
+  getPosition() {
+    return this.begin + this.idx + 1;
+  }
+
+  gotoPosition(pos) {
+    if (this.uids.length === 0) {
+      return;
+    }
+
+    let idx = pos - this.begin - 1;
+
+    if (idx < 0) {
+      idx = 0;
+    }
+    if (idx > this.uids.length - 1) {
+      idx = this.uids.length - 1;
+    }
+
+    this.goto(this.uids[idx]);
+  }
+
+  onFirstUnit() {
+    return this.idx <= 0; // -1 or 0
+  }
+
+  onLastUnit() {
+    return this.idx === this.uids.length - 1;
+  }
+
+  firstUnitInVicinity(vicinity) {
+    return Math.max(0, this.idx - vicinity);
+  }
+
+  lastUnitInVicinity(vicinity) {
+    return Math.min(this.uids.length - 1, this.idx + vicinity);
+  }
+
+  visibleUnitsBefore() {
+    // TODO: instead of just reporting unis to render,
+    // can we use a callback whenever the visible units change?
+    const start = this.firstUnitInVicinity(this.visibleRowsBefore);
+    const end = this.idx - 1;
+    return this.uids.slice(start, end + 1);
+  }
+
+  visibleUnitsAfter() {
+    // TODO: instead of just reporting unis to render,
+    // can we use a callback whenever the visible units change?
+    const start = this.idx + 1;
+    const end = this.lastUnitInVicinity(this.visibleRowsAfter);
+    return this.uids.slice(start, end + 1);
+  }
+
+  prefetchUnits() {
+    const start = this.firstUnitInVicinity(this.visibleRowsBefore + this.prefetchRows);
+    const end = this.lastUnitInVicinity(this.visibleRowsAfter + this.prefetchRows);
+    const uids = this.uids.slice(start, end + 1);
+    const originalUid = this.uid;
+
+    const filteredUids = [];
+
+    for (let i = 0; i < uids.length; i++) {
+      const uid = uids[i];
+      if (!this.units[uid]) {
+        filteredUids.push(uid);
+      }
+    }
+
+    if (!filteredUids.length) {
+      return; // nothing to do
+    }
+
+    UnitAPI.fetchViewRows({ uids: filteredUids }).then(
+      (data) => {
+        let hasNewData;
+
+        this.ready = true;
+
+        if (!data.rows) {
+          return;
+        }
+
+        Object.keys(data.rows).forEach((uid) => {
+          if (this.units[uid]) {
+            return; // unit was already loaded
+          }
+          hasNewData = true;
+          this.units[uid] = new Unit(data.rows[uid]);
+        });
+
+        if (hasNewData) {
+          // TODO: fire this only if the rows in the viewport have changed
+          this.handleViewRowsChange(originalUid);
+        }
+      },
+      this.error
+    );
+  }
+
+  cleanupUnusedUnits() {
+    const start1 = 0;
```

Comment:
I think these are pretty explanatory from the code, since the scope of the function is pretty small. There are two ranges (start1, end1) and (start2, end2) that are calculated to cleanup units that go out of the scope from both sides of the list. (start1, end1) refers to the units above and (start2, end2) reference the range of units below the edit row. So we can probably use something like `aboveStart`, `aboveEnd`, `belowStart` and `belowEnd` if you feel that this is necessary.